Comorphisms/SuleCFOL2TPTP.hs

```diff
@@ -0,0 +1,682 @@
+              name = NameString $ mkSimpleId nameString
+
+              predicateResult = predicateOfSort $ opRes opType
+              predicates = map predicateOfSort $ opArgs opType
+              variables =
+                map (\ i -> variableOfVar $ mkSimpleId $ "X" ++ show i)
+                  [1 .. length (opArgs opType)]
+              function = functionOfOp opName
+
+              functionAntecedent =
+                FOFUF_logic $ FOFLF_binary $ FOFBF_assoc $ FOFBA_and $
+                map (\ (v, p) ->
+                      FOFUF_atomic $ FOFAT_plain $ FOFPAF_predicate p
+                      [FOFT_function $ FOFFT_plain $ FOFPT_constant $ v]) $
+                  zip variables predicates
+              functionConsequent = FOFUF_atomic $ FOFAT_plain $
+                FOFPAF_predicate predicateResult
+                [FOFT_function $ FOFFT_plain $ FOFPT_functor function $
+                  map FOFT_variable variables]
+
+              unitaryFormulaConstant = FOFUF_atomic $ FOFAT_plain $
+                FOFPAF_predicate predicateResult
+                [FOFT_function $ FOFFT_plain $ FOFPT_constant $ function]
+
+              unitaryFormulaFunction = FOFUF_quantified $
+                FOF_quantified_formula ForAll variables $
+                FOFUF_logic $ FOFLF_binary $ FOFBF_nonassoc $
+                FOF_binary_nonassoc
+                  TAS.Implication functionAntecedent functionConsequent
+
+              formula =
+                if null variables
+                then unitaryFormulaConstant
+                else unitaryFormulaFunction
+              sentence = fofUnitaryFormulaToAxiom name formula
+          in  makeNamed nameString sentence
+
+sortOfX :: TAS.Variable -> SORT -> FOF_unitary_formula
+sortOfX var sort = FOFUF_atomic $ FOFAT_plain $
+  FOFPAF_predicate (predicateOfSort sort) [FOFT_variable var]
+
+functionOfOp :: OP_NAME -> TAS.TPTP_functor
+functionOfOp opName =
+  let functionName = "op_" ++ toAlphaNum (show opName)
+  in  mkSimpleId functionName
+
+predicateOfPred :: PRED_NAME -> TAS.Predicate
+predicateOfPred predName =
+  let predicateName = "pred_" ++ toAlphaNum (show predName)
+  in  mkSimpleId predicateName
+
+predicateOfSort :: SORT -> TAS.Predicate
+predicateOfSort sort =
+  let predicateName = "sort_" ++ toAlphaNum (show sort)
+  in  mkSimpleId predicateName
+
+variableOfVar :: VAR -> TAS.Variable
+variableOfVar var =
+  let varName = "VAR_" ++ toAlphaNum (show var)
+  in  mkSimpleId varName
+
+toAlphaNum :: String -> String
+toAlphaNum = concatMap toAlphaNumC
+  where
+    toAlphaNumC :: Char -> String
+    toAlphaNumC c = case c of
+      '+' -> "PLUS"
+      '-' -> "MINUS"
+      '/' -> "SLASH"
+      '\\' -> "BACKSLASH"
+      '%' -> "PERCENT"
+      '<' -> "OPEN"
+      '>' -> "CLOSE"
+      '~' -> "TILDE"
+      '=' -> "EQ"
+      '*' -> "STAR"
+      '\'' -> "PRIME"
+      '\"' -> "QUOTE"
+      '_' -> "_"
+      _ -> if isAlphaNum c then [c] else 'U' : showHex (ord c) ""
```

Comment:
I would prefer to replace a space by an underscore, not by "U20". Note that there anyway is the risk of clashes with existing names.